File: src/tools/pmd-cpd/eval-repos/synthetic/java/HeavyDuplication.java
/**
 * Java file with heavy duplication - multiple large duplicated blocks.
 */
package synthetic;

import java.util.*;

public class HeavyDuplication {

    public static String generateUserReport(List<Map<String, Object>> users) {
        List<String> lines = new ArrayList<>();
        lines.add("=".repeat(60));
        lines.add("USER REPORT");
        lines.add("=".repeat(60));
        lines.add("");

        for (Map<String, Object> item : users) {
            lines.add("ID: " + item.getOrDefault("id", "N/A"));
            lines.add("Name: " + item.getOrDefault("name", "Unknown"));
            lines.add("Email: " + item.getOrDefault("email", "N/A"));
            lines.add("Status: " + item.getOrDefault("status", "active"));
            lines.add("Created: " + item.getOrDefault("createdAt", "Unknown"));
            lines.add("-".repeat(40));
        }

        lines.add("");
        lines.add("Total records: " + users.size());
        lines.add("=".repeat(60));
        return String.join("\n", lines);
    }

    public static String generateAdminReport(List<Map<String, Object>> admins) {
        List<String> lines = new ArrayList<>();
        lines.add("=".repeat(60));
        lines.add("ADMIN REPORT");
        lines.add("=".repeat(60));
        lines.add("");

        for (Map<String, Object> item : admins) {
            lines.add("ID: " + item.getOrDefault("id", "N/A"));
            lines.add("Name: " + item.getOrDefault("name", "Unknown"));
            lines.add("Email: " + item.getOrDefault("email", "N/A"));
            lines.add("Status: " + item.getOrDefault("status", "active"));
            lines.add("Created: " + item.getOrDefault("createdAt", "Unknown"));
            lines.add("-".repeat(40));
        }

        lines.add("");
        lines.add("Total records: " + admins.size());
        lines.add("=".repeat(60));
        return String.join("\n", lines);
    }

    public static String generateGuestReport(List<Map<String, Object>> guests) {
        List<String> lines = new ArrayList<>();
        lines.add("=".repeat(60));
        lines.add("GUEST REPORT");
        lines.add("=".repeat(60));
        lines.add("");

        for (Map<String, Object> item : guests) {
            lines.add("ID: " + item.getOrDefault("id", "N/A"));
            lines.add("Name: " + item.getOrDefault("name", "Unknown"));
            lines.add("Email: " + item.getOrDefault("email", "N/A"));
            lines.add("Status: " + item.getOrDefault("status", "active"));
            lines.add("Created: " + item.getOrDefault("createdAt", "Unknown"));
            lines.add("-".repeat(40));
        }

        lines.add("");
        lines.add("Total records: " + guests.size());
        lines.add("=".repeat(60));
        return String.join("\n", lines);
    }

    public static List<String> validateUserInput(Map<String, Object> data) {
        List<String> errors = new ArrayList<>();
        if (data.get("name") == null) errors.add("Name is required");
        if (data.get("email") == null) errors.add("Email is required");
        String email = (String) data.getOrDefault("email", "");
        if (!email.contains("@")) errors.add("Invalid email format");
        if (data.get("password") == null) errors.add("Password is required");
        String password = (String) data.getOrDefault("password", "");
        if (password.length() < 8) errors.add("Password must be at least 8 characters");
        if (data.get("age") == null) errors.add("Age is required");
        Integer age = (Integer) data.getOrDefault("age", 0);
        if (age < 18) errors.add("Must be at least 18 years old");
        return errors;
    }

    public static List<String> validateAdminInput(Map<String, Object> data) {
        List<String> errors = new ArrayList<>();
        if (data.get("name") == null) errors.add("Name is required");
        if (data.get("email") == null) errors.add("Email is required");
        String email = (String) data.getOrDefault("email", "");
        if (!email.contains("@")) errors.add("Invalid email format");
        if (data.get("password") == null) errors.add("Password is required");
        String password = (String) data.getOrDefault("password", "");
        if (password.length() < 8) errors.add("Password must be at least 8 characters");
        if (data.get("age") == null) errors.add("Age is required");
        Integer age = (Integer) data.getOrDefault("age", 0);
        if (age < 18) errors.add("Must be at least 18 years old");
        return errors;
    }
}

File: src/tools/pmd-cpd/eval-repos/synthetic/go/heavy_duplication.go
// Package synthetic contains Go files for CPD testing with heavy duplication.
package synthetic

import (
	"fmt"
	"strings"
)

// ReportItem represents an item in a report.
type ReportItem struct {
	ID        string
	Name      string
	Email     string
	Status    string
	CreatedAt string
}

// GenerateUserReport creates a formatted report for users.
func GenerateUserReport(users []ReportItem) string {
	var lines []string
	lines = append(lines, strings.Repeat("=", 60))
	lines = append(lines, "USER REPORT")
	lines = append(lines, strings.Repeat("=", 60))
	lines = append(lines, "")

	for _, item := range users {
		id := item.ID
		if id == "" {
			id = "N/A"
		}
		name := item.Name
		if name == "" {
			name = "Unknown"
		}
		email := item.Email
		if email == "" {
			email = "N/A"
		}
		status := item.Status
		if status == "" {
			status = "active"
		}
		createdAt := item.CreatedAt
		if createdAt == "" {
			createdAt = "Unknown"
		}
		lines = append(lines, fmt.Sprintf("ID: %s", id))
		lines = append(lines, fmt.Sprintf("Name: %s", name))
		lines = append(lines, fmt.Sprintf("Email: %s", email))
		lines = append(lines, fmt.Sprintf("Status: %s", status))
		lines = append(lines, fmt.Sprintf("Created: %s", createdAt))
		lines = append(lines, strings.Repeat("-", 40))
	}

	lines = append(lines, "")
	lines = append(lines, fmt.Sprintf("Total records: %d", len(users)))
	lines = append(lines, strings.Repeat("=", 60))
	return strings.Join(lines, "\n")
}

// GenerateAdminReport creates a formatted report for admins - duplicated structure.
func GenerateAdminReport(admins []ReportItem) string {
	var lines []string
	lines = append(lines, strings.Repeat("=", 60))
	lines = append(lines, "ADMIN REPORT")
	lines = append(lines, strings.Repeat("=", 60))
	lines = append(lines, "")

	for _, item := range admins {
		id := item.ID
		if id == "" {
			id = "N/A"
		}
		name := item.Name
		if name == "" {
			name = "Unknown"
		}
		email := item.Email
		if email == "" {
			email = "N/A"
		}
		status := item.Status
		if status == "" {
			status = "active"
		}
		createdAt := item.CreatedAt
		if createdAt == "" {
			createdAt = "Unknown"
		}
		lines = append(lines, fmt.Sprintf("ID: %s", id))
		lines = append(lines, fmt.Sprintf("Name: %s", name))
		lines = append(lines, fmt.Sprintf("Email: %s", email))
		lines = append(lines, fmt.Sprintf("Status: %s", status))
		lines = append(lines, fmt.Sprintf("Created: %s", createdAt))
		lines = append(lines, strings.Repeat("-", 40))
	}

	lines = append(lines, "")
	lines = append(lines, fmt.Sprintf("Total records: %d", len(admins)))
	lines = append(lines, strings.Repeat("=", 60))
	return strings.Join(lines, "\n")
}

// GenerateGuestReport creates a formatted report for guests - duplicated structure.
func GenerateGuestReport(guests []ReportItem) string {
	var lines []string
	lines = append(lines, strings.Repeat("=", 60))
	lines = append(lines, "GUEST REPORT")
	lines = append(lines, strings.Repeat("=", 60))
	lines = append(lines, "")

	for _, item := range guests {
		id := item.ID
		if id == "" {
			id = "N/A"
		}
		name := item.Name
		if name == "" {
			name = "Unknown"
		}
		email := item.Email
		if email == "" {
			email = "N/A"
		}
		status := item.Status
		if status == "" {
			status = "active"
		}
		createdAt := item.CreatedAt
		if createdAt == "" {
			createdAt = "Unknown"
		}
		lines = append(lines, fmt.Sprintf("ID: %s", id))
		lines = append(lines, fmt.Sprintf("Name: %s", name))
		lines = append(lines, fmt.Sprintf("Email: %s", email))
		lines = append(lines, fmt.Sprintf("Status: %s", status))
		lines = append(lines, fmt.Sprintf("Created: %s", createdAt))
		lines = append(lines, strings.Repeat("-", 40))
	}

	lines = append(lines, "")
	lines = append(lines, fmt.Sprintf("Total records: %d", len(guests)))
	lines = append(lines, strings.Repeat("=", 60))
	return strings.Join(lines, "\n")
}

// InputData represents input data for validation.
type InputData struct {
	Name     string
	Email    string
	Password string
	Age      int
}

// ValidateUserInput validates user input data.
func ValidateUserInput(data InputData) []string {
	var errors []string
	if data.Name == "" {
		errors = append(errors, "Name is required")
	}
	if data.Email == "" {
		errors = append(errors, "Email is required")
	}
	if !strings.Contains(data.Email, "@") {
		errors = append(errors, "Invalid email format")
	}
	if data.Password == "" {
		errors = append(errors, "Password is required")
	}
	if len(data.Password) < 8 {
		errors = append(errors, "Password must be at least 8 characters")
	}
	if data.Age == 0 {
		errors = append(errors, "Age is required")
	}
	if data.Age < 18 {
		errors = append(errors, "Must be at least 18 years old")
	}
	return errors
}

// ValidateAdminInput validates admin input data - duplicated validation.
func ValidateAdminInput(data InputData) []string {
	var errors []string
	if data.Name == "" {
		errors = append(errors, "Name is required")
	}
	if data.Email == "" {
		errors = append(errors, "Email is required")
	}
	if !strings.Contains(data.Email, "@") {
		errors = append(errors, "Invalid email format")
	}
	if data.Password == "" {
		errors = append(errors, "Password is required")
	}
	if len(data.Password) < 8 {
		errors = append(errors, "Password must be at least 8 characters")
	}
	if data.Age == 0 {
		errors = append(errors, "Age is required")
	}
	if data.Age < 18 {
		errors = append(errors, "Must be at least 18 years old")
	}
	return errors
}
